### ISSUE TITLE
Provide a way to configure which packages are copied and which not

### ISSUE THREAD

**Issue by jorgemoralespou** [carvel-dev/imgpkg#282]
**Describe the problem/challenge you have**
When relocating (imgpkg copy) a packagerepository that is old and has too many old versions into a new cluster/datacenter, we might know that old versions are never going to be used, so we want to blacklist them from copying over, so that size of images in destination repository and process of copying is improved.

**Describe the solution you'd like**
I would like to have a Config file that could have a blacklist/whitelist of packages that would be copied so that in the destination only the copied versions would exist and only the images for the specified package versions are copied.

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

**Comment by cppforlife** [comment]
(... need to think how to bridge the gap here between imgpkg concepts and packages == imgpkg does not know what's a package intentionally.)

**Comment by joaopapereira** [comment]
@jorgemoralespou is your request somewhat of a different approach to this epic https://github.com/vmware-tanzu/carvel-imgpkg/issues/120 ?

**Comment by jorgemoralespou** [comment]
If the epic "pulling" also covers "copying" probably not.

**Comment by joaopapereira** [comment]
I am trying to think through what are the impacts of this change.
In your scenario really sounds compelling the need for this feature, nevertheless this could cause some strange state between registries.

Let's assume this scenario you have a Bundle A that you want to copy to an air-gapped environment but we only want to copy 2 images that are part of that bundle.
Now, this bundle will exist in 2 places differently even if they are they have the same SHA. Since you have this bundle in an air-gapped environment you might not be able to get the missing images if you need them.

On the other hand, I believe this could be interesting for very big Bundles that contain common images, like an OS or something that the user knows it will exist in the destination register. Of course, this will somewhat complicate the logic of `imgpkg` because now it would have to find images not present in the bundle in the destination registry.


I foresee some possible implementation problems due to the fact that images might not be present when we expected them to be. Do you know of any other particular scenario where this feature could be helpful @jorgemoralespou

**Comment by cppforlife** [comment]
> Let's assume this scenario you have a Bundle A that you want to copy to an air-gapped environment but we only want to copy 2 images that are part of that bundle.

i think the main use case is not to copy set of images, but rather copy subset of bundles (e.g. package bundles). so result would be (in packaging case at least) to be able to install some subset of packages.

**Comment by jorgemoralespou** [comment]
Yeah, this is really related to kapp-controller package repository relocation, where relocating a repository with old version would be really resource intensive.
Kapp-controller on one hand recommends to only add to image repositories, very much like any packaging system, but then relocation means you need to copy all of it, and it might eventually be years of images. I would expect a way that imgpkg can filter out a set of bundles based on some policies. Maybe kapp-controller or other tool will need to create that list based on some querying logic to an existing repository. This is where I see that Carvel is about composable tools that collaborate. (e.g .imgpkg/images.yaml file created by imgpkg and used by kapp-controller as well)

**Comment by joaopapereira** [comment]
After talking a little bit more about this issue, I think that the concept of a not complete Bundle might not be a major issue.

This could open a window for people to only copy images that they know are needed in the destination Registry, especially in an environment without internet access. This could save a lot of bandwidth when transporting the Images.

Questions that we need to think about:
1. Since we assume that maybe not all images would be available when doing a pull a Bundle we might need to go ahead and update the ImagesLock even if 1 Image is no longer present (Maybe the new UX proposed in #284 would help to display this information to the user)
2. When copying a partial Bundle should we try to get the Images from their original repository if no "filtering" is provided or should we somehow keep the Bundle partial?
3. What would we want the filter the be on? Labels on images or Repositories or Image names?

**Comment by cppforlife** [comment]
> What would we want the filter the be on? Labels on images or Repositories or Image names?

+ and how does that connect to e2e experience with pkgrepo+pkgs.

**Comment by joaopapereira** [comment]
Going to mark this issue as a discussion since it looks like there are some open questions
